### ISSUE TITLE
Threading issues when using with EF Core

### ISSUE THREAD

**Issue by avin-kavish** [graphql-dotnet/graphql-dotnet#1726]
I'm getting this error when using EF Core with graphql dotnet,

>  Microsoft.EntityFrameworkCore.Query[10100]
>       An exception occurred while iterating over the results of a query for context type 'PilotRoster.Data.RosterContext'.
>       System.InvalidOperationException: A second operation started on this context before a previous operation completed. This is usually caused by different threads using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
>          at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()

**Comment by sungam3r** [comment]
See #1310

**Comment by avin-kavish** [comment]
There's few different solutions being discussed in that thread. My ideal solution would be to create a scope per parallel thread launched by graphql-dotnet, is that possible?

**Comment by sungam3r** [comment]
Yes. @Shane32 may help.

**Comment by Shane32** [comment]
Your scenario is covered perfectly with PR #1571 .   With it, there is simply another method on the field builder for creating scoped resolvers (`ResolveAsyncScoped`).  It relies on introducing `IServiceProvider` to the `IResolveFieldContext` interface, which does not yet exist.  So until it is merged (if ever), there is no easy way to write an extension method to create a scope for the necessary services.  Instead, you will need to do something like this:

```csharp
public class MyGraphType : ObjectGraphType<MySourceType>
    private readonly IServiceProvider _rootServiceProvider;

    public MyGraphType(IServiceProvider rootServiceProvider) {
        _rootServiceProvider = rootServiceProvider;
        ...
        FieldAsync(.... , resolve: MyFieldResolver);
    }

    private async Task<MyObject> MyFieldResolver(IResolveFieldContext<MySourceType> context) {
        using var scope = context.RequestServices.CreateScope();
        var scopedServiceProvider = scope.ServiceProvider;

        //your code here, for example:
        var db = scopedServiceProvider.GetRequiredService<MyObjectContext>();
        // be sure to await the result so the scope is not disposed before the data has been returned
        return await db.People.Select(x => x.Deleted == false).ToListAsync(context.CancellationToken);
    }
}
```

The `using` statement will dispose of all objects created within the scope once the resolver has finished executing.  Note that you can use any existing service provider (e.g. `HttpContext.RequestServices`) to call `CreateScope()`, and it will function the same with MS's DI implementation.  In other words, MS does not support nested scopes, so every new scope will always be independent.  In the sample above, I am pulling the root service provider via the constructor of the graph type, as it is easiest to demonstrate.

**Comment by avin-kavish** [comment]
Thank @Shane32. I will have to go with option 1 as the code is already written similarly. Will this work seamlessly with `DataLoader`?

**Comment by avin-kavish** [comment]
But to address the larger issue of consuming non-thread-safe scoped services within a parrallel execution context, we need to find a better solution.

**Comment by Shane32** [comment]
Are you making other asynchronous calls from your resolvers, or is the dbcontext mainly the only asynchronous code in your graphs?  Do your graphs use asynchronous methods or synchronous methods to call the dbcontext?

**Comment by avin-kavish** [comment]
DbContext is the only asynchronous code (for this project). Graphs use asynchronous methods.

**Comment by Shane32** [comment]
I'm guessing that your scenario describes my scenario pretty closely.  The business logic layers, in my case, talk to each other, and an object created by a dbcontext in one class (e.g. FooService) might be passed to another object (e.g. BarService) and from there SaveChanges is called on the dbcontext created by BarService.  If your code works similarly, it is essential that FooService and BarService are created by the same dbcontext, or else SaveChanges will not recognize the changes to the object created by the other dbcontext.  This means that providing a service scope is not optional, but necessary to the business logic layers.

**Comment by avin-kavish** [comment]
Yes, I'm doing the same thing. But that's only for mutations (which are executed in sequence). For queries, we don't necessarily need the same db context.

**Comment by Shane32** [comment]
> DbContext is the only asynchronous code (for this project). Graphs use asynchronous methods.

In that case, a multithreaded execution strategy is pointless if you synchronize access to the dbcontext, because the execution strategy always runs synchronous resolvers synchronously (even if they return a Task).  It's only when an asynchronous Task is returned (and still pending completion) that it will move on to executing another resolver.

So if you are considering using a lock for the calls to dbcontext, then you might as well just use a SerialExecutionStrategy.

Either that, or create service scopes when calling the business layer.

**Comment by Shane32** [comment]
> Data loaders

@avin-kavish The data loader is horribly broken in the current version.  I wouldn't use it for anything.  See PR #1511 for a list of bugs and fixes available with the new data loader implementation I've written.  The new data loader uses the same syntax as the old one, for the most part, so if it ever gets pulled in, it should change very little code (if any).

In answer to your question, the same rules of thumb apply to the data loader.  For example, the load function of the data loader (where it calls the business execution layer) would need to create a scope if you are using a `ParallelExecutionStrategy`

**Comment by Shane32** [comment]
One of the biggest problems with the data loader is that it currently doesn't work at all with the `SerialExecutionStrategy`.  This has been fixed in the revised data loader PR.  So you are not tied to the `ParallelExecutionStrategy` if you want to use data loaders.

**Comment by avin-kavish** [comment]
> The data loader is horribly broken in the current version. 

Yeah, haha, I'm starting to have regrets about using this library. This one is production ready though, [https://github.com/MichalLytek/type-graphql](url) but it's in typescript.

`SerialExecutionStrategy`  will cause n+1 issue though?

**Comment by Shane32** [comment]
Well, if they would pull the data loader PR in, and if you write your code to extensively use data loaders, you should have no problems at all.  (Because the new data lower doesn't cause n+1 issue with SerialExecutionStrategy)  Yes, it will still run serially, but this won't even matter much because typically the graphs requested are nested more than they are parallel.  And in which case the lower tiers' data loaders can't run until the upper tiers run anyway.  So I find that it's not really a problem to use a serial execution strategy.  Does that make sense?

So this might not be able to run in parallel:
```
{
  manufacturers {
    id
    name
  }
  categories {
    id
    name
  }
}
```
But this won't run any different:
```
{
  products {
    manufacturers {
      id
      pictures {
        url
      }
    }
  }
}
```
I'm sure some queries, like for a product, will run faster with parallel, but then, those are often the queries that don't matter if they are a tad slower.  And it all depends on how long your data layer is taking to respond.  With Azure database and Azure web services, the db calls are often < 1ms.  So serial/parallel isn't making or breaking the speed here.  n+1 is the real thing you want to avoid, as you've said.

> I'm starting to have regrets about using this library

The library is written quite well, except for the data loaders, and they do not want to add a `RequestServices` property to `IServiceProvider`, which really places limits on the DI-based extensions that can be written.  Fortunately the data loader piece can be extracted as a separate library - I could post it as a separate nuget package and it would work fine.  But the guts of it has been speed optimized quite well, and it covers quite a large portion of the graphql specifications; more than I use for sure.

**Comment by avin-kavish** [comment]
> The library is written quite well

For one, conventions based object and graph types would be nice, just like ef core. At the very least, attribute-based.

**Comment by Shane32** [comment]
Yeah, that's what I was working on in with PR #1376.  There's also another project https://github.com/graphql-dotnet/conventions which I have not used, but may help with that also.

I wanted more control over my graphs, so it is not really tied to the data layer that much, so I don't use conventions for my projects.

My goal was to have graphql requests pull from the database based on the requested graph, so I started using https://github.com/SimonCropp/GraphQL.EntityFramework -- but it is not able to pull only the requested fields.  So I wrote a library similar to it from scratch, which works well.  Now any graph query pulls only the requested fields, including child graphs (where the child is a single entity) from the database (e.g. "manufacturer").  Any child graphs with multiple entries (e.g. "pictures") use data loaders to load, and again pull information based on the requested graph.

**Comment by avin-kavish** [comment]
Yep, I like that conventions style.

**Comment by Shane32** [comment]
> I didn't really get how the serial execution strategy works.

With the updated data loaders, it works like this:

query:
```
{
  products {
    manufacturers {
      pictures {
    }
  }
}
```

This happens:
1. The root query executes a request against the database and retrieves the product list from the database
2. The ProductGraph.Manufacturers field is executed 20 times.  Each time it queues another request for a manufacturer into a data loader based on the id of the product.  The resolvers return an object of type `IDataLoaderResult<Manufacturer>`
3. The execution strategy executes every field it can (above), until it has no other fields to execute.  Then it looks for all the `IDataLoaderResult<>` results it has retrieved.  It will then execute them serially.
4. Upon executing the first `IDataLoaderResult<>`, the manufacturer data loader will load the data for all 20 products.  This is one call, and if there was 10 manufacturers for each product, it would return 200 results.
5. The remainder of the `IDataLoaderResult<>` calls just return cached data for the other 199 results.
6. Repeat step 2 - 200 times for ManufacturerGraph.Pictures.
7. Repeat step 3
8. Repeat step 4 - data loader for pictures now loads pictures for 200 manufacturers.  If there was 5 pictures/manufacturer, it would return 1000 rows from the database.
9. Repeat step 5 - for another 999 rows.

Make sense?  Only 3 db calls.

**Comment by Shane32** [comment]
There are variations of the data loader depending on the database relationship (one-to-one, or one-to-many, or many-to-many)

**Comment by avin-kavish** [comment]
Right, so the updated loader would pretty much batch all entities in the same level. But what about the current loader?

**Comment by Shane32** [comment]
The current loader does something pretty similar, but only works with `ParallelExecutionStrategy`.  So your data loader loading code (and any other resolvers) must be thread safe.  It has other bugs, including some unique deadlock situations, all listed in the PR.

**Comment by avin-kavish** [comment]
Yeah that's what I was saying, so if I switch to Serial now, and since your PR is not merged, I would end up in a n+1 scenario.

**Comment by Shane32** [comment]
There's one important thing to realize when using data loaders (of any kind) with SQL.  The generated SQL will generally be something like:
```sql
SELECT * FROM Manufacturers WHERE ProductId IN (2,3,593,3,994)
```
And it's impossible to parameterize the inputs.  So SQL can't cache the execution strategy.  Just something to be aware of.

**Comment by Shane32** [comment]
> Yeah that's what I was saying, so if I switch to Serial now, and since your PR is not merged, I would end up in a n+1 scenario.

Yup.  But copying the relevant cs files into your project might be easier than trying to make everything thread safe.

**Comment by avin-kavish** [comment]
Hmm.... I'm reading through it now, where does the magic happen? All I know about data loading is that in node js, it's done by collecting all requests across one cycle of the event loop. But since there's no event loop in C#, how is the batch period decided?

**Comment by Shane32** [comment]
Read here: https://github.com/graphql-dotnet/graphql-dotnet/pull/1511/files#diff-ec064dd49f3de7af42c88307d6f7350d

1. Line 20 runs a loop until it processes all outstanding resolvers.  If any are data loader results, they get queued at line 33
2. Line 45 executes all outstanding data loaders.
3. Line 18 loops steps 1 & 2.

**Comment by Shane32** [comment]
In some scenarios (when loaders are on different levels), loaders do not get batched - for example:
```
{
  manufacturers {
    pictures { url }
  }
  categories {
    products {
      pictures { url }
    }
  }
}
```
Then because picture is listed twice but at different levels, they will end up getting batched separately. Well, unless products was not using a data loader, then it would effectively be at the same level.  Just follow the 3 steps above and you can predict what will be batched and what won't.

**Comment by avin-kavish** [comment]
About that, this makes the execution strategy data-loader aware right? Where as previously it wasn't.

But what I meant was without taking the execution strategy into account, say an app calls `LoadAsync` 5 times, how does the loader know when to batch and send the request? Is it after the 3rd call, 4th, or 5th?

**Comment by Shane32** [comment]
> About that, this makes the execution strategy data-loader aware right? Where as previously it wasn't.

Correct.  This is why the data loader was broken before - it could not make intelligent decisions about when to execute the data loaders and instead relied on `OnBeforeExecutionStepAwaitedAsync`.

> say an app calls LoadAsync 5 times, how does the loader know when to batch and send the request

It will not during any of the LoadAsync calls.  Rather, once step 2 is hit above, for each data loader node it calls `CompleteDataLoaderNodeAsync` - see line 48 https://github.com/graphql-dotnet/graphql-dotnet/pull/1511/files#diff-ec064dd49f3de7af42c88307d6f7350dR48 .  This calls `dataLoaderResult.GetResultAsync` which executes the data loader and loads all queued values.  If the data loader has already been run, it merely returns the cached value.

https://github.com/graphql-dotnet/graphql-dotnet/pull/1511/files#diff-450013550b30dd4b3c7f10a518ab0a6fR240

**Comment by Shane32** [comment]
Once `dataLoaderResult.GetResultAsync` is run, the dataloader is "reset" for future calls to `LoadAsync` so they get batched separately.  And when those `DataLoaderResult<>.GetResultAsync` methods get called, the process is repeated.  So there are no bugs if the same data loader is used multiple times in one graphql call.

Plus, if caching is enabled (optional), any results previously used in this request will be returned rather than re-queued to the database.  (part of the existing functionality in the library)

**Comment by Shane32** [comment]
I would have named the functions a bit differently, but the naming is there to match the existing method signatures of the data loader.  So `LoadAsync` just queues a pending result, and `FetchAsync` is where you write your code to load data from the database.  (if I remember correctly)

**Comment by avin-kavish** [comment]
In principle data loader is supposed to operate anywhere, not just in a graphql executor. Isn't there an implementation that could work without the recieving party needing to know that it's loading through a data loader?

Can't we hide this implementation detail inside the data loader itself and expose a regular task to the execution strategy?

**Comment by avin-kavish** [comment]
I came up with this pattern that makes some design sense. The only change to the business logic layer with this way is that it becomes multi-threading aware and loads data accordingly. Like so,

```csharp

    public class SparesContext : IdentityDbContext<User, Role, int>
    {
        
        public DbSet<Part> Parts { get; set; }
        
        public SparesContext(DbContextOptions<SparesContext> options): base(options)
        {
            
        }
    }

    public class SparesContextFactory
    {
        private readonly DbContextOptions<SparesContext> _options;

        public SparesContextFactory(DbContextOptions<SparesContext> options)
        {
            _options = options;
        }

        public SparesContext Produce()
        {
            return new SparesContext(_options);
        }
    }
    
    public class ParallelQueryContext
    {
        private readonly SparesContextFactory _factory;

        public ParallelQueryContext(SparesContextFactory factory)
        {
            _factory = factory;
        }

        public DbSet<Part> Parts => _factory.Produce().Parts;
    }

    public class PartsService
    {
        private readonly ParallelQueryContext _parallelQueryContext;

        public PartsService(ParallelQueryContext parallelQueryContext)
        {
            _parallelQueryContext = parallelQueryContext;
        }
        
        public async Task<IEnumerable<Part>> Find()
        {
            return await _parallelQueryContext.Parts.Where(x => x.Name == "Something").ToListAsync();
        }
    }

```

**Comment by avin-kavish** [comment]
New and improved version using generics. Also got rid of the parallel query context, felt like it didn't really serve a meaningful purpose.

```csharp
    public class SparesContext : IdentityDbContext<User, Role, int>
    {
        public DbSet<Part> Parts { get; set; }
        
        public SparesContext(DbContextOptions<SparesContext> options): base(options)
        {
            
        }
    }

    public interface IContextFactory<out T>
    {
        T Produce();
    }

    public class SparesContextFactory: IContextFactory<SparesContext>
    {
        private readonly DbContextOptions<SparesContext> _options;

        public SparesContextFactory(DbContextOptions<SparesContext> options)
        {
            _options = options;
        }

        public SparesContext Produce()
        {
            return new SparesContext(_options);
        }
    }

    public interface IParallelQueryRunner<TContext>
    {
        Task<T> RunInParallel<T>(Func<TContext, Task<T>> func);
    }

    public class ParallelQueryRunner<TContext> : IParallelQueryRunner<TContext> where TContext : IAsyncDisposable
    {
        private readonly IContextFactory<TContext> _factory;

        public ParallelQueryRunner(IContextFactory<TContext> factory)
        {
            _factory = factory;
        }

        public async Task<T> RunInParallel<T>(Func<TContext, Task<T>> func)
        {
            await using var context = _factory.Produce();
            return await func(context);
        }
    }

    public class PartService
    {
        private readonly IParallelQueryRunner<SparesContext> _parallelQueryRunner;

        public PartService(IParallelQueryRunner<SparesContext> parallelQueryRunner)
        {
            _parallelQueryRunner = parallelQueryRunner;
        }
        
        public async Task<IEnumerable<Part>> Find()
        {
            return await _parallelQueryRunner.RunInParallel(async context => 
                await context.Parts.OrderByDescending(p => p.DateCreated).ToListAsync()
            );
        }
    }

```

**Comment by Shane32** [comment]
If this works for your business logic layers, or if you want to spend that much time rewriting the necessary code, that's great.  I fear that you'll run into problems with untracked entities if your database layers pass tracked objects between them.  It would seem to me to be much simpler to create a scope in your queries and leave your database layers alone:

```csharp
public static class ScopeExtensions
{
    public static async Task<TReturn> RunScopedAsync<TSource, TReturn>(this IResolveFieldContext<TSource> context, Func<IResolveFieldContext<TSource>, IServiceProvider, Task<TReturn>> func)
    {
        using (var scope = ((IServiceProvider)context.Schema).CreateScope()) {
            return await func(context, scope.ServiceProvider);
        }
    }
    public static FieldBuilder<TSource, TReturn> ResolveScopedAsync<TSource, TReturn>(this FieldBuilder<TSource, TReturn> builder, Func<IResolveFieldContext<TSource>, IServiceProvider, Task<TReturn>> func)
    {
        return builder.ResolveAsync(context => context.RunScopedAsync(func));
    }
}

public class Query : ObjectGraphType
{
    public Query()
    {
        Field<ProductGraph, IEnumerable<Product>>()
            .Name("Products")
            .Description("Return all products")
            .ResolveScopedAsync(async (context, serviceProvider) => {
                return serviceProvider.GetRequiredService<ProductService>().SearchProductsAsync();
            });
    }
}
```

I really don't think it takes that much time to spin up another service scope.  I've tested it in the past, but I can't remember now exactly what my findings were.

I'll be interested to know what route you take and what you find once you start implementing it.

**Comment by avin-kavish** [comment]
Hmm... would it be possible to inject into `ResolveScopedAsync` without having to use `GetRequiredService` ?

**Comment by Shane32** [comment]
> Don't we have a package like `GraphQL.Net.AspNetCore.DI` that supports DI in resolvers somehow?

Not that I'm aware of.  @sungam3r has a library called SteroidsDI that may help:

https://github.com/sungam3r/SteroidsDI

> Hmm... would it be possible to inject into `ResolveScopedAsync` without having to use `GetRequiredService` ?

You could possibly write a field builder extension function that accepts the service type as a generic type parameter and calls `GetRequiredService` within the extension function.  But then you need to type all three generic type parameters, and it gets messy.

```csharp
        Field<ProductGraph, IEnumerable<Product>>("Products")
            .Description("Return all products")
            .ResolveScopedAsync<ProductGraph, IEnumerable<Product>, ProductService> (async (context, service) => {
                return service.SearchProductsAsync();
            });
```

If you like this concept, you can probably figure out a way to make it cleaner.

**Comment by avin-kavish** [comment]
Hmm... yeah, since most of the time I only need one service, I wouldn't mind this. But I could do something like,

```csharp
      public class ProductsResolver : IResolver
      {
             public ProductsResolver(/* DI */)
             {
                 
             }

             public IEnumerable<Product> ExecuteAsync()
             {
                   /* logic */
             }
       }

       Field<ProductGraph, IEnumerable<Product>>("Products")
            .Description("Return all products")
            .ResolveScopedAsync<ProductGraph, ProductsResolver>();
```

if I need many services to be injected, but then I would have a class-per-resolver which is just too many classes. It would be better to model many resolvers in a single class with perhaps parameter injection using `[FromServices]`, which brings me back to your PR #1376 

Can't we have it as a seperate package similar to `Graphql.Conventions`? Even in a different repository with a different namespace is fine. It may even be more appropriate that way because it's opinionated in the ASP.Net style.

**Comment by Shane32** [comment]
The changes in PR #1376 is already in a separate project; you can just compile it and use it alongside the base graphql-dotnet nuget package.  That's how I use it.  I could post it as a nuget package if you would find that useful.

**Comment by avin-kavish** [comment]
Yes, please, thank you.

**Comment by fenomeno83** [comment]
Here a full implementation on .net core 3.1 with scoped problem solved
https://github.com/fenomeno83/graphql-dotnet-globalization-demo

**Comment by avin-kavish** [comment]
@fenomeno83 your implementation doesn't seem to create a scope each time a resolver is executed, it seems to be using the scope of the http context, i.e the request. Our discussion is that, that will not suffice because the graphql spec specifies resolvers can be executed in parallel within the context of a single request, which leads to threading issues with EF Core.

**Comment by fenomeno83** [comment]
You're right.
Here
https://github.com/fenomeno83/graphql-dotnet-globalization-demo
i've added another approach that solves parallel execution problem; it is based on factory pattern, so you can instantiate dbcontext explicitly (in this case there is an example that shows how pass dbcontext between the various methods that works also alone)

**Comment by fenomeno83** [comment]
@avin-kavish I've revised another time code here
https://github.com/fenomeno83/graphql-dotnet-globalization-demo

I've used @Shane32 approach, that in my opinion is the best.
We create a scope at resolver level, so you haven't issues with entity framewok.
I've created some extensions method and service method so IServiceProvider is not directly exposed

**Comment by sungam3r** [comment]
There is a non-zero probability that work on DI issues will be continued soon.

**Comment by Shane32** [comment]
Close @sungam3r due to #2261 being merged?

**Comment by sungam3r** [comment]
OK. Do we have enough documentation now? I mean exactly for the Entity Framework usecase. This is often asked and will continue to ask.

**Comment by Shane32** [comment]
I think so.  Both 3.x and 4.x have samples explicitly for EF Core within the documentation page for Dependency Injection.  (The 4.x samples are updated for #2261 )

**Comment by Shane32** [comment]
@avin-kavish @fenomeno83 Within v4 we have more support for DI included as a separate nuget package.  Please review the documentation as soon as v4 is released.  If you have other questions, please re-open or open a new issue.  Thanks!
